refactor(list-images): consolidate image listing

KnifeyMoloko opened the pull request:
This PR introduces `list-images` to SCT commands that allows to list machine images from both AWS and GCE. The command options allow to filter the results by: cloud backend, architecture, region (for AWS results only), branch.  Additionally it also allows to list the images by version. This combines the previously available `list-ami-branch`, `list-gce-branch`, `list_gce_images_versions` and `list_aws_images_versions`  into one command.

Trello task: https://trello.com/c/DHZ7mE3u

Asciinema screencast with sample usage:
https://asciinema.org/a/v2LuoLh2o93ekRcOFDmsLXljU

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

fruch commented on sdcm/utils/common.py:
```diff
@@ -1526,14 +1528,64 @@ def get_branched_ami(scylla_version: str, region_name: str, arch: AwsArchType =
+
+
+def create_pretty_table(rows: list[str], field_names: list[str]) -> PrettyTable:
+    tbl = PrettyTable(field_names=field_names)
```
please add `align="l"`

KnifeyMoloko replied:
Fixed (added `align="l"`)

fruch commented on sct.py:
```diff
@@ -505,38 +504,6 @@ def list_ami_versions(region: str, arch: AwsArchType):
-    if ":" not in version:
-        version += ":all"
-
-    tbl = PrettyTable(field_names=["Name", "ImageId", "CreationDate", "BuildId", "Test Status"], align="l")
```
Seem like you are throwing away the "Test Status".

it a feature we have that can report a longevity result back to the used AMI, as a tag.

we didn't used it, and if yanking it, maybe we should yank all of it.

KnifeyMoloko replied:
Yup, I removed it as it seems we don't use these values. What do you mean by yanking "all of it"?

KnifeyMoloko replied:
We have `tag_ami_with_result` in `tester.py`, but it is getting the result from the event summary. From what I can tell there is no connection between the code removed here and the actual tagging method.

fruch commented on sdcm/utils/common.py:
```diff
@@ -1526,14 +1528,64 @@ def get_branched_ami(scylla_version: str, region_name: str, arch: AwsArchType =
     images = [image for image in images if not image.name.startswith('debug-image')]
 
-    assert images, f"AMIs for {scylla_version=} with {arch} architecture not found in {region_name}"
+    if not images:
```
again don't demote the assert into a log print, this function is being called also from other places which are no via the commandline, we do want it to fail.

fruch commented on sdcm/utils/common.py:
```diff
@@ -1560,7 +1615,9 @@ def get_branched_gce_images(scylla_version: str, project: str = SCYLLA_GCE_IMAGE
     )
 
-    assert images, f"GCE images for {scylla_version=} not found"
+    if not images:
```
keep the assert, catch it on the cli function if you want different output

KnifeyMoloko replied:
Fixed (reverted to an assert and handled in the calling function)

fgelcer reviewed (commented):
other than a naming question, everything else LGTM

fgelcer commented on sct.py:
```diff
@@ -70,7 +68,8 @@
     list_resources_docker,
     search_test_id_in_latest,
-    list_parallel_timelines_report_urls
+    list_parallel_timelines_report_urls, get_ami_images, get_gce_images,
```
aren't these out of order? like, isn't it better to have them in alphabetical order, and one command by line?

KnifeyMoloko replied:
Not sure why PyCharm decided that's the way to place the imports. Either way, I fixed them back to alphabetical order.

fgelcer commented on sct.py:
```diff
@@ -547,25 +514,41 @@ def list_gce_images_versions():
-    \n\n[VERSION] is a branch version to look for, ex. 'branch-2019.1:latest', 'branch-3.1:all'""")
-@click.argument("version", type=str, default="branch-3.1:all")
-def list_gce_images_branch(version):
+@cli.command('list-images', help="List machine images")
```
where? i still see it as it was during my comment

KnifeyMoloko replied:
If you mean the command (i.e. `list-images`) it's not there. You think it should be something like `list-images-branch`? As for the function calls, it's all `branch` and no `version` there, both for the `list_images` function and the functions called inside it.

fgelcer replied:
yes, i think the command should be `list-branch-images` or something like that, to differentiate from `list-release-images` (or `list-version-images`)
like it was before, but in this case, `image` is the unified command for all cloud backends (until Azure will be added to it too)

fgelcer commented on sct.py:
```diff
@@ -487,85 +486,81 @@ def list_resources(ctx, user, test_id, get_all, get_all_running, verbose):
               help="architecture of the AMI (default: x86_64)")
-@click.argument('version', type=str, default='branch-3.1:all')
-def list_ami_branch(region: str, arch: AwsArchType, version: str):
+def list_images(backend: str, branch: str, version: bool, region: str, arch: AwsArchType):
```
version should be also `str` no?

fgelcer commented on sct.py:
```diff
@@ -487,85 +486,81 @@ def list_resources(ctx, user, test_id, get_all, get_all_running, verbose):
+              help="Branch to query images for. Defaults to 'master:latest'")
+@click.option('-v', '--version',
+              type=str,
+              is_flag=True,
```
how do you send a specific version if this is a flag? IIUC a flag is a `boolean`

KnifeyMoloko replied:
The original command did not filter on a version argument, so it made sense to make it a flag. That being said I changed the command somewhat to allow filtering by versions.

fgelcer reviewed (commented):
2 comments, but the calls to these functions in the `groovy` files must be updated as well, or they use already the call to the base functions (that haven't changed)?

fgelcer commented on sct.py:
```diff
@@ -487,85 +486,82 @@ def list_resources(ctx, user, test_id, get_all, get_all_running, verbose):
-        ])
-    click.echo(tbl.get_string(title="Scylla GCE images branch versions"))
+    branch_fields = ["Backend", "Name", "ImageId", "CreationDate", "BuildId", "Arch", "ScyllaVersion"]
+    version_fields = ["Backend", "Name", "ImageId", "CreationDate"]
```
shouldn't it have also `Arch` and `BuildId` in here?

fruch replied:
FYI, this is a bug, I've raised it with scylla-pkg:
https://github.com/scylladb/scylla-cluster-tests/issues/5041

we should have labels across all images.

fgelcer replied:
> FYI, this is a bug, I've raised it with scylla-pkg: #5041
> 
> we should have labels across all images.

@fruch , this issue 5041 isn't an issue to scylla-pkg... 

@KnifeyMoloko , please add a comment to the `version_fields` mentioning the issue @fruch mentioned (https://github.com/scylladb/scylla-pkg/issues/2995) so we will know that we can expand it once the issue is fixed

fruch replied:
Well the bug I've paste is kind of related, if we'll want azure support :)

KnifeyMoloko replied:
TODO comment added. I also created a task in Trello, so we don't loose sight of this one (https://trello.com/c/UoVmGEmw)

fgelcer reviewed (commented):
other than 2 small comments, it looks great, and seems pretty easy to enhance it to add Azure images as well, so well done so far

fruch commented on sct.py:
```diff
@@ -487,85 +488,82 @@ def list_resources(ctx, user, test_id, get_all, get_all_running, verbose):
-    \n\n[VERSION] is a branch version to look for, ex. 'branch-2019.1:latest', 'branch-3.1:all'""")
+@cli.command('list-images', help="List machine images")
+@click.option('-b', '--backend',
+              type=click.Choice(["aws", "gce", "azure"]),
```
This command not yet support azure, don't think I should be mentioned here

KnifeyMoloko replied:
Rather than removing it, I added another `case` to communicate that it's not yet implemented, as an additional reminder of the intent here.

KnifeyMoloko replied:
Fixed (added `@cloud_provider_option`)

fruch commented on sct.py:
```diff
@@ -487,85 +488,82 @@ def list_resources(ctx, user, test_id, get_all, get_all_running, verbose):
+              help="List images by version. Use '-v all' for all versions. "
+                   "OSS format: <4.3> Enterprise format: <enterprise-2021.1>")
 @click.option('-r', '--region',
-              type=CloudRegion(cloud_provider="aws"),
```
Why didn't you kept the CloudRegion ? It can validate the input of region based on the backend

KnifeyMoloko replied:
We need to provide the `cloud_provider` argument in the signature though. IIUC there is no way to "share" the `backend` argument. It made sense to use it for backend-specific commands, but I don't see a way to use it for the "all-in-one" command here.

fruch replied:
no we don't need to provide `cloud_provider` see `create_runner_instance`, it can read the provider from click's context.

KnifeyMoloko replied:
Ok, that makes sense. I've added `CloudRegion`.

fgelcer reviewed (dismissed):
other than a comment (to add a comment in the code, with the `scylla-pkg` issue), LGTM

fruch reviewed (dismissed):
LGTM

fgelcer reviewed (dismissed):
LGTM

fruch reviewed (approved):
LGTM

fruch commented:
@KnifeyMoloko, seems like you've lost a recent fix, and the table of AWS doesn't show build-id correctly
```diff
diff --git a/sdcm/utils/common.py b/sdcm/utils/common.py
index 30c94327..bb755d75 100644
--- a/sdcm/utils/common.py
+++ b/sdcm/utils/common.py
@@ -1558,7 +1558,7 @@ def get_ami_images(branch: str, region: str, arch: AwsArchType) -> list:
     for ami in amis:
         tags = {i['Key']: i['Value'] for i in ami.tags}
         rows.append(["AWS", ami.name, ami.image_id, ami.creation_date, tags.get(
-            'build-id'), tags.get('arch'), tags.get('ScyllaVersion')])
+            'build-id', tags.get('build_id'))[:6], tags.get('arch'), tags.get('ScyllaVersion')])

     return rows
```

fruch commented:
@KnifeyMoloko 

this commands fails:
```
❯ ./sct.py list-images -v 5.0.0

...

botocore.exceptions.NoRegionError: You must specify a region.
```

1) seems like the default now AWS is the default ( cause of cloud provider decorator), and region needs to be defined.
2) the default matches when no provider, doesn't work when no region is passed in the command line.

something like the following helped with having different default and make it not required
```diff
diff --git a/sct.py b/sct.py
index 4ce6946f..0c79bc7d 100755
--- a/sct.py
+++ b/sct.py
@@ -27,7 +27,7 @@ import uuid
 import pprint
 from concurrent.futures import ProcessPoolExecutor
 from pathlib import Path
-from functools import partial
+from functools import partial, wraps
 from typing import Optional
 from uuid import UUID

@@ -137,14 +137,19 @@ def add_file_logger(level: int = logging.DEBUG) -> None:
     LOGGER.addHandler(handler)


-cloud_provider_option = click.option(
-    "-c", "--cloud-provider",
-    required=True,
-    type=click.Choice(choices=SUPPORTED_CLOUDS, case_sensitive=False),
-    default=DEFAULT_CLOUD,
-    is_eager=True,
-    help="Cloud provider",
-)
+def cloud_provider_option(function=None, default: str | None = DEFAULT_CLOUD, required: bool = True):
+    def actual_decorator(f):
+        return click.option(
+            "-c", "--cloud-provider",
+            required=required,
+            type=click.Choice(choices=SUPPORTED_CLOUDS, case_sensitive=False),
+            default=default,
+            is_eager=True,
+            help="Cloud provider",
+        )(f)
+    if function:
+        return actual_decorator(function)
+    return actual_decorator
```

fruch reviewed (dismissed):
1) fix build-id output
2) make provider and region defaults work/validated out of the box

KnifeyMoloko commented:
> 1. fix build-id output
> 2. make provider and region defaults work/validated out of the box

@fruch 
1. Fixed with applying your suggestion.
2. Fixed - you can now provide a cloud provider or not (and it will not default to "aws"), regions are validated in the functions called by the `list_images` itself.

fgelcer reviewed (dismissed):
LGTM

fgelcer commented:
@KnifeyMoloko fix conflicts

KnifeyMoloko commented:
> @KnifeyMoloko fix conflicts

Conflicts resolved.

fgelcer reviewed (approved):
LGTM

fruch commented:
> > 1. fix build-id output
> > 2. make provider and region defaults work/validated out of the box
> 
> @fruch
> 
> 1. Fixed with applying your suggestion.
> 2. Fixed - you can now provide a cloud provider or not (and it will not default to "aws"), regions are validated in the functions called by the `list_images` itself.

this command still doesn't work, since there's no default region.
```bash
./sct.py list-images -v 5.0.0
```

and even worse:

```bash
❯ ./sct.py list-images -v 5.0.0 --region us-east-1
Usage: sct.py list-images [OPTIONS]
Try 'sct.py list-images --help' for help.

Error: Invalid value for '-r' / '--region': unknown cloud provider: None
```

so the whole default of not setting the cloud provider, kind of moot, and it's a code path we can never get to.


or we remove the code path that show images from multiple providers, and provide sane defaults to arguments.
or we make sure we set defaults in that code path

KnifeyMoloko commented:
> 



> > > 1. fix build-id output
> > > 2. make provider and region defaults work/validated out of the box
> > 
> > 
> > @fruch
> > 
> > 1. Fixed with applying your suggestion.
> > 2. Fixed - you can now provide a cloud provider or not (and it will not default to "aws"), regions are validated in the functions called by the `list_images` itself.
> 
> this command still doesn't work, since there's no default region.
> 
> ```shell
> ./sct.py list-images -v 5.0.0
> ```
> 
> and even worse:
> 
> ```shell
> ❯ ./sct.py list-images -v 5.0.0 --region us-east-1
> Usage: sct.py list-images [OPTIONS]
> Try 'sct.py list-images --help' for help.
> 
> Error: Invalid value for '-r' / '--region': unknown cloud provider: None
> ```
> 
> so the whole default of not setting the cloud provider, kind of moot, and it's a code path we can never get to.
> 
> or we remove the code path that show images from multiple providers, and provide sane defaults to arguments. or we make sure we set defaults in that code path

This is strange. When I run the same command locally, I'm getting this output (as expected, from both implemented cloud provider backends):

```
+----------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                             Machine Images by version                                                              |
+---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
| Backend | Name           | ImageId                                                                                 | CreationDate                  |
+---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
| AWS     | ScyllaDB 5.0.0 | ami-050b81d683f061fbd                                                                   | 2022-06-30T16:41:11.000Z      |
| GCE     | scylla-5-0-0   | https://www.googleapis.com/compute/v1/projects/scylla-images/global/images/scylla-5-0-0 | 2022-06-30T09:55:44.269-07:00 |
+---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+

```

fruch commented:
> > 
> 
> > > > 1. fix build-id output
> > > > 2. make provider and region defaults work/validated out of the box
> > > 
> > > 
> > > @fruch
> > > 
> > > 1. Fixed with applying your suggestion.
> > > 2. Fixed - you can now provide a cloud provider or not (and it will not default to "aws"), regions are validated in the functions called by the `list_images` itself.
> > 
> > 
> > this command still doesn't work, since there's no default region.
> > ```shell
> > ./sct.py list-images -v 5.0.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and even worse:
> > ```shell
> > ❯ ./sct.py list-images -v 5.0.0 --region us-east-1
> > Usage: sct.py list-images [OPTIONS]
> > Try 'sct.py list-images --help' for help.
> > 
> > Error: Invalid value for '-r' / '--region': unknown cloud provider: None
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so the whole default of not setting the cloud provider, kind of moot, and it's a code path we can never get to.
> > or we remove the code path that show images from multiple providers, and provide sane defaults to arguments. or we make sure we set defaults in that code path
> 
> @fruch This is strange. When I run the same command locally, I'm getting this output (as expected, from both implemented cloud provider backends):
> 
> ```
> +----------------------------------------------------------------------------------------------------------------------------------------------------+
> |                                                             Machine Images by version                                                              |
> +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> | Backend | Name           | ImageId                                                                                 | CreationDate                  |
> +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> | AWS     | ScyllaDB 5.0.0 | ami-050b81d683f061fbd                                                                   | 2022-06-30T16:41:11.000Z      |
> | GCE     | scylla-5-0-0   | https://www.googleapis.com/compute/v1/projects/scylla-images/global/images/scylla-5-0-0 | 2022-06-30T09:55:44.269-07:00 |
> +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> ```
> 
> As for querying with `region` but without `cloud-provider` it kinda doesn't make sense, same as mixing up `branch` and `version`. So it looks like I'll remove the multiprovider route here and make "aws" the default for cloud-provider.


seem like the code assumes there a default region configured, and it shouldn't assume that:

```bash
❯ cat ~/.aws/config
[default]
```

we had issue with running hydra command in various places, we shouldn't assume default region was set by the user.

KnifeyMoloko commented:
> > > 
> > 
> > 
> > > > > 1. fix build-id output
> > > > > 2. make provider and region defaults work/validated out of the box
> > > > 
> > > > 
> > > > @fruch
> > > > 
> > > > 1. Fixed with applying your suggestion.
> > > > 2. Fixed - you can now provide a cloud provider or not (and it will not default to "aws"), regions are validated in the functions called by the `list_images` itself.
> > > 
> > > 
> > > this command still doesn't work, since there's no default region.
> > > ```shell
> > > ./sct.py list-images -v 5.0.0
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and even worse:
> > > ```shell
> > > ❯ ./sct.py list-images -v 5.0.0 --region us-east-1
> > > Usage: sct.py list-images [OPTIONS]
> > > Try 'sct.py list-images --help' for help.
> > > 
> > > Error: Invalid value for '-r' / '--region': unknown cloud provider: None
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > so the whole default of not setting the cloud provider, kind of moot, and it's a code path we can never get to.
> > > or we remove the code path that show images from multiple providers, and provide sane defaults to arguments. or we make sure we set defaults in that code path
> > 
> > 
> > @fruch This is strange. When I run the same command locally, I'm getting this output (as expected, from both implemented cloud provider backends):
> > ```
> > +----------------------------------------------------------------------------------------------------------------------------------------------------+
> > |                                                             Machine Images by version                                                              |
> > +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> > | Backend | Name           | ImageId                                                                                 | CreationDate                  |
> > +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> > | AWS     | ScyllaDB 5.0.0 | ami-050b81d683f061fbd                                                                   | 2022-06-30T16:41:11.000Z      |
> > | GCE     | scylla-5-0-0   | https://www.googleapis.com/compute/v1/projects/scylla-images/global/images/scylla-5-0-0 | 2022-06-30T09:55:44.269-07:00 |
> > +---------+----------------+-----------------------------------------------------------------------------------------+-------------------------------+
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As for querying with `region` but without `cloud-provider` it kinda doesn't make sense, same as mixing up `branch` and `version`. So it looks like I'll remove the multiprovider route here and make "aws" the default for cloud-provider.
> 
> seem like the code assumes there a default region configured, and it shouldn't assume that:
> 
> ```shell
> ❯ cat ~/.aws/config
> [default]
> ```
> 
> we had issue with running hydra command in various places, we shouldn't assume default region was set by the user.

Ah, makes sense. I made the following changes:
1. Removed the multiprovider route - we now default to "aws" as cloud-provider value.
2. Added default to region ("eu-west-1").
3. Updated the help messages to reflect the defaulrs.

fgelcer reviewed (approved):
LGTM

soyacz reviewed (approved):
LGTM
